src/proofnetwork/xjdf/data/resource.cpp

```diff
@@ -0,0 +1,215 @@
+{
+    ResourceSP resource;
+    QVector<PartSP> parts;
+    AmountPoolSP amountPool;
+    QString id;
+    ResourceOrientation orientation = ResourceOrientation::Rotate0Orientation;
+
+    if (reader.isStartElement() && reader.name() == QStringLiteral("Resource")) {
+        auto attributes = reader.attributes();
+        if (attributes.hasAttribute(QStringLiteral("ID")))
+            id = attributes.value(QStringLiteral("ID")).toString();
+        if (attributes.hasAttribute(QStringLiteral("Orientation")))
+            orientation = resourceOrientationFromString(attributes.value(QStringLiteral("Orientation")).toString());
+        reader.readNextStartElement();
+
+        while (!reader.atEnd() && !reader.hasError()) {
+            if (reader.isStartElement() && reader.name() == QStringLiteral("AmountPool")) {
+                amountPool = AmountPool::fromXJdf(reader, document);
+            } else if (reader.isStartElement() && reader.name() == QStringLiteral("Part")) {
+                auto part = Part::fromXJdf(reader, document);
+                if (part)
+                    parts << part;
+            } else if (reader.isStartElement()) {
+                auto creator = resourceCreator(reader.name().toString());
+                if (creator) {
+                    resource = creator(reader, document);
+                    reader.readNext();
+                    continue;
+                }
+            } else if (reader.isEndElement() && reader.name() == QStringLiteral("Resource")) {
+                break;
+            }
+            reader.readNext();
+        }
+    }
+
+    if (resource) {
+        resource->setId(id);
+        resource->setOrientation(orientation);
+        resource->setParts(parts);
+        resource->setAmountPool(amountPool);
+    }
+
+    return resource;
+}
+
+Resource::Resource(const QString &id) : Resource(*new ResourcePrivate, id)
+{}
+
+Resource::Resource(ResourcePrivate &dd, const QString &id) : XJdfAbstractNode(dd)
+{
+    setId(id);
+}
+
+void Resource::addResourceCreator(const QString &name,
+                                  std::function<ResourceSP(QXmlStreamReader &, const XJdfDocumentSP &)> &&creator)
+{
+    if (!ResourcePrivate::creators)
+        ResourcePrivate::creators =
+            new QMap<QString, std::function<ResourceSP(QXmlStreamReader &, const XJdfDocumentSP &)>>();
+    if (!(*ResourcePrivate::creators).contains(name))
+        (*ResourcePrivate::creators)[name] = creator;
+}
+
+void Resource::updateSelf(const Proof::NetworkDataEntitySP &other)
+{
+    ResourceSP castedOther = qSharedPointerCast<Resource>(other);
+    setId(castedOther->id());
+    setParts(castedOther->parts());
+    setOrientation(castedOther->orientation());
+    setAmountPool(castedOther->amountPool());
+
+    XJdfAbstractNode::updateSelf(other);
+}
+std::function<ResourceSP(QXmlStreamReader &, const XJdfDocumentSP &)> &Resource::resourceCreator(const QString &name)
```

Comment:
I don't see any reason to have this function as protected interface of public class. I'd suggest to keep it as internal function at least for now. 
Also it will modify creators map in case of unknown names and it is harder to protect it with any sort of mutex if such need arises. I'd suggest pass here both reader and document and return ResourceSP. Something like this:
```
ResourceSP createResource(name, reader, document) {
  auto creator = creators.find(name);
  return creator == creators.end() ? ResourceSP() : creator.value()(reader, document);
}
```